src/ops/base.rs

```diff
@@ -67,8 +249,46 @@ impl Ingredient for Base {
                 rs: Records,
                 _: &DomainNodes,
                 state: &StateMap)
-                -> Records {
-        rs.into_iter()
+                -> Option<Records> {
+        // Write incoming records to log before processing them if we are a durable node.
+        let records_to_return;
+        let mut copy_rs = rs.clone();
```

Comment:
Why is this `clone` necessary? It seems pretty costly, especially because it's not just executed for durable nodes.